### PR TITLE
Teach elmer to chat with the Snowflake Cortex Analyst

### PR DESCRIPTION
**Pull request by atheriel**
This commit introduces `chat_cortex()`, a new provider that allows the user to chat with the [Snowflake Cortex Analyst][0].

The Analyst has a bespoke -- and underdocumented -- API, so a fair amount of this is custom handling determined by trial and error. The major distinction of Cortex Analyst is that it returns a structured response that can contain text, SQL, and "suggestions" -- rather than simply a block of text. `elmer` is somewhat equipped to handle this already, though I had to make changes in a few places to accommodate it.

Under the hood, the Analyst is powered by Llama and Mistral models, but the caller doesn't specify the model at all. Nor do they provide a system prompt -- instead, the caller points to something that Snowflake calls a ["semantic model"][1], which is a YAML file with a known specification.

The Analyst also doesn't support multi-turn, though I expect it will in the future (and the error message you get when trying indicates as much). There is also no support for tool calling.

I've implemented (and verified) both authentication methods that the API supports, namely Snowflake OAuth and what they term "key-pair" authentication, which uses JWTs signed by an RSA key. Key-pair authentication can be especially useful because it should allow us to run tests against Cortex in CI.

Workbench-managed credentials are also supported.

Unit tests are included. Integration tests also run if there are ambient Snowflake credentials available.

### Demo

Demo Shiny chat app for testing, using the semantic model from Snowflake's tutorial:

```r
library(shiny)
library(shinychat)
library(elmer)

# You'll need to set up key-pair auth for testing on desktop.
# See: https://docs.snowflake.com/en/developer-guide/snowflake-rest-api/authentication#using-key-pair-authentication
Sys.setenv(
  "SNOWFLAKE_ACCOUNT" = "<snip>",
  "SNOWFLAKE_USER" = "<snip>",
  "SNOWFLAKE_PRIVATE_KEY" = "rsa_key.p8"
)

stage <- "CORTEX_ANALYST_DEMO.REVENUE_TIMESERIES.RAW_DATA"
model_file <- "revenue_timeseries.yaml"

ui <- bslib::page_fluid(
  chat_ui("chat", "What questions can I ask?")
)

server <- function(input, output, session) {
  chat <- chat_cortex(
    model_file = glue::glue("@{stage}/{model_file}")
  )

  chat_append("chat", glue::glue(
    "❄️ Hi, I'm the [Snowflake Cortex Analyst](https://docs.snowflake.com/en/user-guide/snowflake-cortex/cortex-analyst)!
I can answer questions about the `{model_file}` data model or generate SQL
queries for you."
  ))

  observeEvent(input$chat_user_input, {
    chat_append("chat", chat$stream_async(input$chat_user_input))
  })
}

shinyApp(ui, server)
```

[0]: https://docs.snowflake.com/en/user-guide/snowflake-cortex/cortex-analyst
[1]: https://docs.snowflake.com/en/user-guide/snowflake-cortex/cortex-analyst/semantic-model-spec

### CODE REVIEW

**Comment by hadley** [comment]
Fixes #56

**Comment by atheriel** [comment]
FWIW, I'm still iterating on their streaming format and figuring out how to make returning structured and/or markdown-formatted text work nicely.

File: R/turns.R [hunk 1]
```diff
@@ -16,7 +19,7 @@ Turn <- new_class(
     text = new_property(
       class = class_character,
       getter = function(self) {
-        is_text <- map_lgl(self@contents, S7_inherits, ContentText)
+        is_text <- map_lgl(self@contents, prop_exists, "text")
```
[the remark this answers is not in the document]
Comment (atheriel) [reply R/turns.R]: That's what I've got now. It does seem better.

File: R/chat.R [hunk 2]
```diff
@@ -366,6 +366,7 @@ print.Chat <- function(x, ...) {
       user = cli::col_blue,
       assistant = cli::col_green,
       system = cli::col_br_white,
+      analyst = cli::col_br_blue,
```
[the remark this answers is not in the document]
Comment (atheriel) [reply R/chat.R]: Sure, happy to follow a convention of conversion.

File: R/api-cortex.R [hunk 3]
```diff
@@ -0,0 +1,459 @@
...
+#' )
+#' chat$chat("What questions can I ask?")
+#' @export
+chat_cortex <- function(account = Sys.getenv("SNOWFLAKE_ACCOUNT"),
+                        credentials = cortex_credentials,
+                        model_spec = NULL,
+                        model_file = NULL,
+                        api_args = list(),
+                        echo = FALSE) {
+  if (!.rlang_check_is_string(account, FALSE, FALSE, FALSE)) {
```
[the remark this answers is not in the document]
Comment (atheriel) [reply R/api-cortex.R]: I can't recall, I've changed it.

File: tests/testthat/test-cortex.R [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    # Tests roundtrip conversion.
+    cortex_message(cortex_message_to_turn(sample_cortex_message)),
+    sample_cortex_message
+  )
+})
+
+test_that("Cortex turn formatting", {
+  turn <- cortex_message_to_turn(sample_cortex_message)
+  expect_snapshot(turn@text)
+  expect_snapshot(format(turn))
```
Comment (hadley) [inline tests/testthat/test-cortex.R:31]: Maybe `cat()` these?

File: tests/testthat/test-cortex.R [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    )
+  )
+  req <- chat_request(p, FALSE, list(turn))
+  expect_snapshot(req)
+  expect_snapshot(req$body$data)
+})
+
+test_that("a simple Cortex chatbot works", {
+  skip_if(
+    Sys.getenv("SNOWFLAKE_ACCOUNT") == "",
```
[the remark this answers is not in the document]
Comment (atheriel) [reply tests/testthat/test-cortex.R:122]: Working on that.

**Comment by hadley** [comment]
Merging just so you don't need to keep dealing with all the merge conflicts that I'm creating. Would definitely love to figure out how to test this locally and on GHA so I don't accidentally break it in the future. (But I'm about to go on vacation until Tuesday evening, so no rush)

**Pull request merged**